sklego/meta.py

```diff
@@ -100,7 +152,69 @@ def __remove_groups_from_x(self, X):
         else:
             return np.delete(X, self.groups, axis=1)
 
-    def fit(self, X, y):
+    def __fit_grouped_estimator(self, X, y, value_columns, group_columns):
+        group_indices = X.groupby(group_columns).indices
+
+        try:
+            grouped_estimations = {
+                group: clone(self.estimator).fit(X.loc[indices, value_columns], y.loc[indices])
+                for group, indices in group_indices.items()
+            }
+        except KeyError:
+            raise KeyError("The indices of y and X should match")
+
+        return grouped_estimations
+
+    def __check_shrinkage_func(self):
+        """Validate the shrinkage function"""
+        group_lengths = [10, 5, 2]
+        expected_shape = np.array(group_lengths).shape
+        try:
+            result = self.shrinkage_function_(group_lengths)
+            if not isinstance(result, np.ndarray):
+                raise ValueError(f"shrinkage_function({group_lengths}) should return an np.ndarray")
+            if result.shape != expected_shape:
+                raise ValueError(f"shrinkage_function({group_lengths}).shape should be {expected_shape}")
+        except Exception as e:
+            raise ValueError(f"Caught an exception while checking the shrinkage function: {str(e)}")
+
+    def __set_shrinkage_function(self):
+        if isinstance(self.shrinkage_function, str):
+            shrink_options = {
+                "constant": constant_shrinkage,
+                "relative": relative_shrinkage,
+                "min_n_obs": min_n_obs_shrinkage,
+            }
+            try:
+                self.shrinkage_function_ = shrink_options.get(self.shrinkage_function)
+            except AttributeError:
+                raise ValueError(f"The specified shrink function {self.shrinkage_function} is not valid, "
+                                 f"choose from {list(shrink_options.keys())}")
+        else:
+            self.shrinkage_function_ = self.shrinkage_function
```

Comment:
why the variables with trailing underscores?  Trailing underscores are used when variables clash with python built-in functions. e.g. `filter_`.

Reply:
It's also sklearn convention for attributes that are learned in `fit`

Reply:
ah! Did not know that. 

Then the question is the `shrinkage_function` "learned" during fit?

Reply:
I moved the initialisation of the `shrinkage_function` to the init because one of Cor's other remarks, so also removed the trailing underscore